agent/consul/gateways/controller_gateways.go

```diff
@@ -74,19 +74,66 @@ func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Req
 // along to either a cleanup function if the entry no longer exists (it's been deleted),
 // or a reconciler if the entry has been updated or created.
 func reconcileEntry[T structs.ControlledConfigEntry](store *state.Store, logger hclog.Logger, ctx context.Context, req controller.Request, reconciler func(ctx context.Context, req controller.Request, store *state.Store, entry T) error, cleaner func(ctx context.Context, req controller.Request, store *state.Store) error) error {
-	_, entry, err := store.ConfigEntry(nil, req.Kind, req.Name, req.Meta)
+	_, rawentry, err := store.ConfigEntry(nil, req.Kind, req.Name, req.Meta)
 	if err != nil {
 		requestLogger(logger, req).Warn("error fetching config entry for reconciliation request", "error", err)
 		return err
 	}
 
-	if entry == nil {
+	if rawentry == nil {
 		return cleaner(ctx, req, store)
 	}
 
+	entry := attemptConfigEntryDeepCopy(rawentry)
+
 	return reconciler(ctx, req, store, entry.(T))
 }
 
+// attemptConfigEntryDeepCopy casts a generic configentry to the appropriate type and deep copies it, if able, or returns
+// the same uncopied config etnry if it cannot
+func attemptConfigEntryDeepCopy(ce structs.ConfigEntry) structs.ConfigEntry {
```

Comment:
I found this to be needed because not all configentries implement DeepCopy, however, I'm not the most familiar with go generics yet and if theres a cleaner way to handle this I'd love to hear it.

Reply:
this LGTM, if we had a way to interface assert against a generic interface type we could maybe clean do an early return or something like that (and if we were on 1.21 I think there's been a bunch of upgrades to the type inference which could make this less verbose but I'm not 100% on that)

Reply:
Can we just assume that all of our controlled config entries (it looks like JWTProvider and APIGateway are the only things missing here?) implement `DeepCopy`? If so I think we can extract these into a private interface rather than do all of the type reflection:

```
type deepCopyable[T any] interface {
  structs.ControlledConfigEntry
  DeepCopy() T
}
```

Here's an example with go 1.20: https://go.dev/play/p/n18XJ-GNFiS?v=goprev

In other places where you aren't using an overloaded config entry (like the gateway-specific reconciliation functions), I'm thinking just call `DeepCopy` directly after doing a cast to the desired type since you know exactly what's going to get returned from the store.

Reply:
do we know they all implement DeepCopy? It doesn't look like APIGateway or JWTProvider implement DeepCopy (we could add them to the generated code so that they do implement it and simplify this a bit)

Reply:
It would be way easier if we could just add it to our controlled config entries yeah, I just didn't want to add them in in case there was a reason why they weren't added to begin with.